Document server architecture and causility/concurrency strategy (#8028)
Scille/parsec-cloud: server/README.md

> `backend` fixture, as this is a very convenient to customize the content of the
> server by directly using the business logic methods (e.g. `backend.vlob.create(...)`).

The web application's main class is obsviously `AsgiApp` (we use
[FastAPI as ASGI framework](https://fastapi.tiangolo.com/), so this class is in
fact a `fastapi.FastAPI`) that is build from a factory that, among other things,
introspect the components to register the HTTP RCP API endpoints they expose.

This is defined in the `parsec/asgi` folder, the HTTP endpoints are:

- RCP API (`parsec/asgi/rpc.py`): this implement the HTTP API used by the Parsec client
  and defined in the crate `libparsec_protocol`.
- Administration API (`parsec/asgi/administration.py`): this is a JSON REST API designed
  for the administration of the server (e.g. create a new organization).
- Redirection (`parsec/asgi/redirect.py`): allow conversion between `https://` -> `parsec://`
  URLs (used in email given `parsec://` URLs are often not displayed correctly by email clients)
- Misc user-facing URLs (`parsec/asgi/__init__.py`), e.g. `index.html`, page not found etc.

> I should be obvious now that the main part of the server is those components that both

Reviewer: ```suggestion
It should be obvious now that the main part of the server is those components that both
```

Author: :+1:

implement the business application logic and the HTTP endpoints for the RPC API.

On top of that those components have two implementations:

- PostgreSQL `parsec/components/postgresql`
- In memory `parsec/components/memory`

While the PostgerSQL implementation is the only one to use in production, the memory
implementation is very convenient:

> - It simplicity allows for fast hack when testing new ideas.

Reviewer: ```suggestion
- Its simplicity allows for fast hack when testing new ideas.
```

Author: :+1:

Any change done by a client that should be visible by the server is provided as a
certificate (i.e. a certificate is an unencrypted document signed by the client's device).

A very simple way to ensure causality and concurrency on certificates would be:

- Have a single big lock in the server that must be taken before any attempt to
  add a certificate (handles concurrency).
- Ensure the new certificate's timestamp is newer than any existing certificate
  (handles causality).

However this is pretty wasteful given most certificates are unrelated (e.g. certif A
add a new user, certif B does a key rotation in a realm). So instead we reduce the
size of the lock by introducing the concept of topic.

Each type of certificate is part of one of the following topics:

- `common`
- `sequester`, which depends on `common`
- `realm`, which depends on `common` and uses the realm ID as discriminant
- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant

So the actual rule used in the server is:

- Have lock in the server for each couple topic + discriminant
- Take the lock for the topic + discriminant couple the certificate we want to add

>   belongs to AND to other topic our topic may depend on (handles concurrency).

Reviewer 2: ```suggestion
  belongs to AND to other topics our topic may depend on (handles concurrency).
```

>   belongs to AND to other topic our topic may depend on (handles concurrency).

Reviewer 2: If I understand correctly, currently all topics depends on `common`, that mean each topic need to lock `common`.

That feels like a _single big lock_, I'm missing something ?

Author: The difference is the `common` lock is taken for read only, and another lock is taken for write.

So for instance we can have concurrent add of certificates in two different realms (both locking the `common` lock for read).

On the other hand, in a big lock architecture the lock would be always taken for write.

Reviewer 2: When the `common` lock is taken in write mode /

Author: `common` lock is taken when adding common certificates, so for instance when adding/revoking a new

This shouldn't have an impact on vlob/block performances given adding common certificates should be rare.

- Ensure the new certificate's timestamp is newer than any existing certificate

>   in the topics we have locked (handles causality).

Reviewer 3: Also, the section does not mention the causality dependencies that exist between certificates and other types of data such as blocks or vlobs.

>   in the topics we have locked (handles causality).

Reviewer: Should we always take the lock on common, as everything depends on it ?

>   in the topics we have locked (handles causality).

Author: Do you think I should add something to stress the importance of this here ?

>   in the topics we have locked (handles causality).

Reviewer: I think so, it' quite important to remember that common is not just another topic

> Note: Also organizations are totally isolated from each other on the server, so
> in a way the organization ID is an aditional discriminant on each certificate topic.

### Blocks

Blocks are encrypted piece of data that belong to a given realm and cannot
be modified.

This makes concurrency & causality pretty simple:

- Each block is considered independent, hence there is no need for lock dedicated to blocks.
- A block is identified by a client-provided UUID, then there is no risk for double-creation
  under concurrency (a trivial unicity check on insertion prevents this).
- Block creation date must be more recent than the last certificate in `common` topic and

>   in the `realm` topic corresponding to the block's realm.

Reviewer 3: This causality check is only treated one way (checking the topic when creating a block). The opposite is not mentioned in the above section.

Reviewer 3: Also this section does not mention locking: is the `common` topic lock taken why the block is uploaded?

> Notes:
>
> - It could be tempting to only lock the `realm` topic (given it's this topic that contains
>   the certificate describing the role of a user in the realm). However a revoked user is also
>   forbidden from adding data, hence making it mandatory to also lock the `common` topic here.
> - Uploading blocks (and vlobs) are the most common task on the server, so we may want to

> >   uptomise it in the future to avoid having to lock two topics.

Reviewer 3: That sounds quite bad. The proposed strategy should already address this problem IMO.

Author: I've run dictionary check and corrected plenty !

Reviewer 3: I didn't mean the syntax, I mean that maybe it's not something to optimize in the future but deal with now in this design document.

>   A possible solution would be to have an optimistic locking approach: first read the
>   last timestamp in the topics, then do the checks, then on the final data insertion check

> >   again the timestamp haven't changed (and if they have we re-do everything from the start).

Reviewer: What would prevent a DoS if new certificates are continuously produced ?

Author: All API where lock is involved are authenticated, hence it is trivial for server administrator to freeze the users doing the DoS using the administration REST API.

Reviewer 2: I understand that's trivial for server admin, but how that's possible on the SaaS ?

Author: I guess the whole point of SAAS is to not having to deal with things such as DoS (as it's the SAAS's administrators job) ;-)

### Vlobs

Vlobs are encrypted metadata that belong to a given realm. Unlike block they can be
modified and hence have a version field that must increase linearly.

> Concurrency & causality is handle is a similar fashion than for blocks.

Reviewer 3: Really? Vlobs require a database write, but blocks require bother a database write and upload to the object storage. I would find very surprising if both used the same approach in the end.

Author: You can think of both vlob and block upload as:

1) fetch info from database
2) do stuff in Python
3) store the vlob/block in the database

In the case of vlob, step 2 is only about checking access rights and vlob existence. On the other hand block also includes the write to the object storage.

However this is irrelevant from the database point of view: in both case step 3 is only allowed if data in step 1 hasn't changed.

Obviously this approach means we can end up with data in the  object storage while the SQL transaction got cancelled.
However this is acceptable:
- Given the block ID is determined client-side, re-trying to upload the block will lead to another write in the object storage at the same location.
- The object storage follows an eventual consistency strategy, hence the new block upload is guaranteed to overwrite the previous one. The only issue would be if the client upload a different block with the same ID between each try, in this case eventual consistency means it is possible for a short period of time to get a previous value when querying the block (but this is purely theoretical since there is no reason for the client to do that !).

> Concurrency & causality is handle is a similar fashion than for blocks.

Reviewer: ```suggestion
Concurrency & causality is handle is a similar fashion as for blocks.
```

Author: That's a subtle one !

### Invitations

During the process of invitation, one of the client (the claimer) doesn't have a device
yet (as it is the whole point of the invitation process !).
Hence using certificates during this step is not possible (since a certificate is signed by a device).

No certificate involvement means:

- Invitations data are totally separated from other data (given other data always relies
  on certificates to do part of there validation).
- Invitation request doesn't contains a client-generated timestamp.

This last point is great from a causility point of view:

- With certificates the client propose a change that occured a given timestamp, then it's
  up to the server to check for causility to know if it should accept it.
- Here the client merely ask the server do to something, then the server check it if it
  possible *now* then do it *now*. No causility can be broken given we only work in the present.

Regarding concurrency, a lock is needed to protect the creation of invitation
(see [example in Data concurrency paragraph](#data-concurrency)).

### Implementation: PostgreSQL

> In PostgreSQL we have multple tables (`common_topic`, `realm_topic` etc.) dediticated to

Reviewer 3: ```suggestion
In PostgreSQL we have multiple tables (`common_topic`, `realm_topic` etc.) dedicated to
```
Many typos haven't been caught by cspell

Author: :+1:

Finally the lock for invitation creation cannot be implemented by regular ROW locking
(given there is no row to lock !), so we use
[an advisory lock instead](https://www.postgresql.org/docs/current/explicit-locking.html).

### Implementation: Memory

Currently memory implementation heavily relies in the fact code is not actually asynchronous.

Indeed, given the data are stored in memory all business logic on them can be done purely
synchronously.

Given we only use a single thread in our application, then a synchronous piece of code
can be seen as atomic (i.e. only `await` change the scheduling of coroutines, not `await`
in an async function means this function will be entirely executed before anything else
can run).

However this is not perfect: adding an `await` to an async function seems harmless, but in
our case it would break the assumption !

On top of that event dispatch is asynchronous event in memory (this is needed to follow
more closely what happen in PostgreSQL when a event may be recieved long after the
request that sent it has finished).
This is not an issue nevetheless given sending event is always the very last operation

> done, at which point the modification in database (and there correspondin checks) have

Reviewer: ```suggestion
done, at which point the modification in database (and their corresponding checks) have
```

Author: :+1:

Any change done by a client that should be visible by the server is provided as a
certificate (i.e. a certificate is an unencrypted document signed by the client's device).

A very simple way to ensure causality and concurrency on certificates would be:

- Have a single big lock in the server that must be taken before any attempt to
  add a certificate (handles concurrency).
- Ensure the new certificate's timestamp is newer than any existing certificate
  (handles causality).

However this is pretty wasteful given most certificates are unrelated (e.g. certif A
add a new user, certif B does a key rotation in a realm). So instead we reduce the
size of the lock by introducing the concept of topic.

Each type of certificate is part of one of the following topics:

- `common`
- `sequester`, which depends on `common`
- `realm`, which depends on `common` and uses the realm ID as discriminant
- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant

So the actual rule used in the server is:

- Have lock in the server for each couple topic + discriminant
- Take the lock for the topic + discriminant couple the certificate we want to add

>   belongs to AND to other topics our topic may depend on (handles concurrency).

Author: That's an interesting point, I've added this information to the sentence.

- Ensure the new certificate's timestamp is newer than any existing certificate
  in the topics we have locked (handles causality).

> Notes:
>
> - Also organizations are totally isolated from each other on the server, so

> >   in a way the organization ID is an additional discriminant on each certificate topic.

Author: We have a `organization` field for (almost) all tables, e.g.

https://github.com/Scille/parsec-cloud/blob/8aa93ec2b2e926d3420ebdd3b2364700f3ee0fed/server/parsec/components/postgresql/migrations/datamodel.sql#L42

> - [As we are going to see](#what-is-locked-by-who), all the other kind of data depends at diverse degrees to
>   the certificates (e.g. blocks are accessed by users that are member of a realm, hence
>   `common` and `realm` topics are involved).

#### `RequireGreaterTimestamp` vs `CertificateBasedActionIdempotentOutcome`

On top of command-specific checks, a server may reject a valid certificate by returning
`RequireGreaterTimestamp` or `CertificateBasedActionIdempotentOutcome`.

`RequireGreaterTimestamp` is simple to understand: a certificate has a timestamp more
recent than the one provided in the command, hence the client should use a timestamp

> strictly superior than the one returned in the error to have a chance to succeed.

Reviewer 2: Should be but I'm in doubt, just want a confirmation :shrug:

> strictly superior than the one returned in the error to have a chance to succeed.

Author: yes, the server checks the timestamp ballpark to prevent this !

  under concurrency (a trivial uniqueness check on insertion prevents this).
- Block creation date must be more recent than the last certificate in `common` topic and
  in the `realm` topic corresponding to the block's realm.

> Notes:
>
> - It could be tempting to only lock the `realm` topic (given it's this topic that contains
>   the certificate describing the role of a user in the realm). However a revoked user is also
>   forbidden from adding data, hence making it mandatory to also lock the `common` topic here.
> - Uploading blocks (and vlobs) are the most common task on the server, so we may want to
>   optimise it in the future to avoid having to lock two topics.
>   A possible solution would be to have an optimistic locking approach: first read the
>   last timestamp in the topics, then do the checks, then on the final data insertion check
>   again the timestamp haven't changed (and if they have we re-do everything from the start).

### Vlobs

Vlobs are encrypted metadata that belong to a given realm. Unlike block they can be
modified and hence have a version field that must increase linearly.

> Concurrency & causality is handle is a similar fashion as for blocks.

Reviewer 2: ```suggestion
Concurrency & causality are handled in a similar fashion as for blocks.
```

Author: done.

API Family    |  command                           | common | realm  | sequester | shamir |
--------------|------------------------------------|--------|--------|-----------|--------|
invited       | ping                               |        |        |           |        |
anonymous     | ping                               |        |        |           |        |
authenticated | ping                               |        |        |           |        |
authenticated | events_listen                      |        |        |           |        |
authenticated | certificate_get                    |        |        |           |        |
anonymous     | organization_bootstrap             |        |        |           |        |
authenticated | invite_list                        |        |        |           |        |
invited       | invite_info                        |        |        |           |        |
authenticated | invite_new_device                  | read   |        |           |        |
authenticated | invite_new_user                    | read   |        |           |        |
authenticated | invite_cancel                      | read   |        |           |        |

> authenticated | invite_1_greeter_wait_peer         |        |        |           |        |

Reviewer: There will be the new invite command to add here after #7761 is merged

Reviewer 2: I would propose to write a script to update the table as it is highly likely that we will forget about it

Author: > I would propose to write a script to update the table as it is highly likely that we will forget about it

There is a script now ! (and it checks more than just that ^^)

> There will be the new invite command

I've added them (because the script forced me :laughing: ), but I have the feeling `invite_greeter_start_greeting_attempt` & `invite_claimer_start_greeting_attempt` also requires some form of manual locking in the style of what is needed when creating an invitation: given the attempt is created by using the user ID as discriminant, but multiple attempts with the same discriminant are allowed in the database (only a single one of them is not removed, but this is not something the database can handle on its own with a unique constraint)

@AureliaDolo  what do you think ?

>
> In the case of vlobs, step 2 is only about checking access rights and vlob existence.
> On the other hand blocks also includes the write to the object storage.
>
> However this is irrelevant from the database point of view: in both case step 3 is only
> allowed if data in step 1 hasn't changed.
>
> Obviously this approach means we can end up with data in the  object storage while the
> SQL transaction got cancelled, however this is acceptable:
>
> - Given the block ID is determined client-side, re-trying to upload the block will lead
>   to another write in the object storage at the same location.
> - The object storage follows an eventual consistency strategy, hence the new block upload
>   is guaranteed to overwrite the previous one.
> - The only issue would be if multiple attempts of the same block ID would be done with
>   different data on each try. In this case eventual consistency means it is possible
>   for a short period of time to get a previous value when querying the block.
>   However this is purely theoretical since block ID is random (hence there should be
>   no concurrency between clients) and a client considers a block as immutable.

> ### Invitations

Reviewer 3: This section should be updated once https://github.com/Scille/parsec-cloud/issues/9302 is treated.

This last point is great from a causality point of view:

- With certificates the client propose a change that occurred at a given timestamp, then it's
  up to the server to check for causality to know if it should accept it.
- Here the client merely ask the server do to something, then the server checks if it is
  possible *now*, then do it *now*. No causality can be broken given we only work in the present.

Regarding concurrency, locks are needed to protect the creation of invitation against:

- Concurrent invitation creation and user creation leading an invitation for an existing user.
  This is handled by simply taking the `common` topic lock.
- Concurrent invitation creations of the same email creating multiple invitations
  (see [example in Data concurrency paragraph](#data-concurrency)). This is handled
  by a dedicated `invitation_creation` advisory lock that must be taken before the
  invitation creation procedure starts any checks involving the invitations.

### What is locked by who ?

The following array recapitulate which topics (and other related locks) each command
needs to lock (and in which mode).

> Topic locks: `common`, `realm`, `sequester`, `shamir`

Reviewer 3: syntax is wrong, should add `\n`

Manual locks: `invitation_creation`
Special checks:

- `last_vlob_timestamp`: Compare the to-be-inserted data's timestamp with the timestamp
  of the most recent vlob.
- `user_revoke_safe`: See [User revocation and concurrent operation involving users](#user-revocation-and-concurrent-operation-involving-users)

<!-- See `misc/check_server_readme.py` script that checks this table is up to date ! -->
<!-- start-table-what-is-lock-by-who -->

> API Family    |  command                               | common | realm | sequester | shamir | invitation_creation | user_revoke_safe | last_vlob_timestamp |

Reviewer 3: Should also add the invitation row lock used by invitation cancel/complete/step etc.)

authenticated | invite_3b_greeter_signify_trust        |        |       |           |        |                     |                  |                     |
authenticated | invite_4_greeter_communicate           |        |       |           |        |                     |                  |                     |
invited       | invite_1_claimer_wait_peer             |        |       |           |        |                     |                  |                     |
invited       | invite_2a_claimer_send_hashed_nonce    |        |       |           |        |                     |                  |                     |
invited       | invite_2b_claimer_send_nonce           |        |       |           |        |                     |                  |                     |
invited       | invite_3a_claimer_signify_trust        |        |       |           |        |                     |                  |                     |
invited       | invite_3b_claimer_wait_peer_trust      |        |       |           |        |                     |                  |                     |

> invited       | invite_4_claimer_communicate           |        |       |           |        |                     |                  |                     |

Reviewer: Should be removed
